torch/distributed/_mesh_layout.py

```diff
@@ -210,3 +211,144 @@ def global_ranks(self, world_size: int) -> list[list[int]]:
             [group_offset + group_rank for group_rank in self.member_ranks()]
             for group_offset in self.complement(world_size).member_ranks()
         ]
+
+    def check_non_overlap(self) -> bool:
+        """
+        Check if the layout has any overlap between the ranks it generates. If there is overlap,
+        we return False, otherwise True.
+
+        Aside from indice 0, indices from each dim of the layout must be non-overlapping.
+
+        Here is how it works:
+        1. Sort dimensions by stride (smallest stride first)
+        2. For each dimension, check if:
+           - It has the same stride as previous dimension (duplicate mapping)
+           - Its stride overlaps with the previous dimension's span
+
+        A dimension's "span" is size * stride, representing the address space it covers.
+
+        Example 1 - Valid (no overlap):
+        Layout: sizes=(2,3), strides=(6,1)
+        - Dim 1: stride=1, span=3*1=3, covers addresses [0,1,2]
+        - Dim 0: stride=6, span=2*6=12, covers addresses [0,6]
+        → No overlap since 6 > 3
+
+        Example 2 - Invalid (overlap):
+        Layout: sizes=(2,3), strides=(2,1)
+        - Dim 1: stride=1, span=3*1=3, covers addresses [0,1,2]
+        - Dim 0: stride=2, span=2*2=4, covers addresses [0,2]
+        → Overlap! stride=2 < span=3, so addresses [0,2] are duplicated
+
+        Returns:
+            bool: True if no overlap exists (valid layout), False if overlap detected
+        """
+        previous_span = -1
+        previous_stride = -1
+        for size, stride in sorted(self.sizes_and_strides, key=lambda x: x[1]):
+            if size == 1:
+                continue
+            if previous_stride == stride or stride < previous_span:
+                return False
```

Comment:
You don't need to test previous_stride == stride. Suppose previous_stride == stride. Then previous_span = size * previous_stride, where size > 1. This then guarantees that previous_span is greater than stride.

Reply:
(this is why I commented about zeros above, technically size can be zero in which case this reasoning doesn't hold, but with zero numel there is never any overlap anyway)